xds/src/main/java/io/grpc/xds/ControlPlaneClient.java

```diff
@@ -308,6 +324,50 @@ final void sendDiscoveryRequest(XdsResourceType<?> type, Collection<String> reso
           respNonces.getOrDefault(type, ""), null);
     }
 
+    private final class AdsStreamEventHandler implements EventHandler<DiscoveryResponse> {
+      @Override
+      public void onReady() {
+        readyHandler();
+      }
+
+      @Override
+      public void onRecvMessage(DiscoveryResponse response) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            XdsResourceType<?> type = fromTypeUrl(response.getTypeUrl());
+            if (logger.isLoggable(XdsLogLevel.DEBUG)) {
+              logger.log(
+                  XdsLogLevel.DEBUG, "Received {0} response:\n{1}", type,
+                  MessagePrinter.print(response));
+            }
+            if (type == null) {
+              logger.log(
+                  XdsLogLevel.WARNING,
+                  "Ignore an unknown type of DiscoveryResponse: {0}",
+                  response.getTypeUrl());
+
+              call.startRecvMessage();
+              return;
+            }
+            handleRpcResponse(type, response.getVersionInfo(), response.getResourcesList(),
```

Comment:
Actually I think this is what the current structure is.  This is the interface between xdsClientImpl and ControlPlaneClient.  The response proto does not have too many interesting fields. Although it might be for sustainable if we define a `Arg` data structure instead of extending the parameter list. But to me it is fine for now and it is internal.